Implied heat transport new diagnostic

mo-abodas opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description

<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->
- Closes #2472 
- Link to documentation: https://esmvaltool--3177.org.readthedocs.build/en/3177/recipes/recipe_iht_toa.html

This recipe implements the analysis documented in [Pearce and Bodas-Salcedo (2023)](https://doi.org/10.1175/JCLI-D-22-0149.1). If CERES-EBAF is used as input dataset then it produces the results in that paper, albeit with small differences. The differences are due to the Edition of the dataset. The paper uses the latest edition (Ed4.1), whereas the recipe uses an older edition available in obs4MIPs (Ed2.7).

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [x] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [x] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [x] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

mo-abodas commented:
Dear @ESMValGroup/science-reviewers this is my first contribution to ESMValTool, so I wonder if you could help me find a science reviewer? Thanks!

alistairsellar commented:
> Dear @ESMValGroup/science-reviewers this is my first contribution to ESMValTool, so I wonder if you could help me find a science reviewer? Thanks!

I could give it a go Alejandro. Probably won't be able to look at it until the end of this week though, so if anyone else would like to review it then please go ahead.

valeriupredoi commented:
Hi @mo-abodas welcome to the Tool, and cheers for your contributions! I can definitely have a look at the code from a technical standpoint, so you can count on me on the technical review side - more frequent scientific reviewers are good folk like @alistairsellar or @axel-lauer or @LisaBock :grin:

valeriupredoi commented on environment.yml:
```diff
@@ -37,6 +37,7 @@ dependencies:
   - natsort
   - nc-time-axis
   - netCDF4
+  - numba
```
it would be great if you didn't use `numba` please - we've had numerous problems with `numba` wrt dependency agreement and environment solving, they don't really offer great support. Sorry to have started with a negative remark :grin:

mo-abodas replied:
> Hi @mo-abodas welcome to the Tool, and cheers for your contributions! I can definitely have a look at the code from a technical standpoint, so you can count on me on the technical review side - more frequent scientific reviewers are good folk like @alistairsellar or @axel-lauer or @LisaBock grin

Thanks @valeriupredoi , @ehogan has also volunteered to do the technical review, but extra eyes are always welcome!

mo-abodas replied:
No worries, I didn't know that, I won't use it. A real shame though, as it gives a massive speed up!

valeriupredoi replied:
Cheers! I'll have a look at the used case and try see how to speed up the code without numba, there's a few other ways, plus, don't just want to be a naysayer without suggesting alternatives 😁

valeriupredoi replied:
actually @mo-abodas I'll take back my reservations wrt numba in the env - we actually bring it in from somewhere (some dependency needs it) so why not add it back as direct dependency, it seems it's been behaving OK lately, see the env in the [github action test](https://github.com/ESMValGroup/ESMValTool/actions/runs/5144236584/jobs/9260276712)

alistairsellar commented on doc/sphinx/source/recipes/recipe_iht_toa.rst:
```diff
@@ -0,0 +1,123 @@
+.. figure::  /recipes/figures/iht_toa/figure4_CERES-EBAF_CERES-EBAF.png
+   :align:   center
+
+   As in Figure 2, but for the implied heat transport associated with
```
Note that figure numbers appear to be dynamic (https://esmvaltool--3177.org.readthedocs.build/en/3177/recipes/recipe_iht_toa.html#example-plots) so Fig 2 actually showing as Fig 39 in this build. Hopefully possible to do a dynamic reference?

mo-abodas replied:
Well spotted @alistairsellar . I don't know how to include a dynamic reference. Happy to do it if someone points me to the right documentation. I think it would be even better if the numbering could be reset to start from 1 for each recipe, that way each recipe would be more self-contained.

alistairsellar replied:
I've not done this for figures or numbered elements before, but if I understand https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html#the-numref-role then I think ```:numref:`fig_iht_toa_2` ``` might work. If not, I guess another option is to replace "As in Figure 2" with a copy/merge of the Fig 2 caption so that the caption is self-contained.

alistairsellar commented on esmvaltool/diag_scripts/iht_toa/poisson_solver.py:
```diff
@@ -0,0 +1,281 @@
+    shp0, shp1 = np.array(cxxx.shape) - 2
+    for j in range(1, shp0 + 1):
+        for i in range(1, shp1 + 1):
+            cxxx[j, i] = m_p[j, i] * \
```
Are the preconditioning steps also described by van der Vorst (1992)? If not, some explanation or another reference for the equations here would be useful.

mo-abodas replied:
Yes, they are. I'm expanding the comments to make that clear.

alistairsellar commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,717 @@
+def symmetry_metric(data, grid):
+    """Calculate symmetry metrics.
+
+    A perfectly symmetrical latitude band gives S=0. As coded, the
```
It is also implicitly assuming a uniform grid spacing in latitude. It feels like it would be safer and more explicit to do cube.extract() for each lat band, which would also remove the multiple of 6 restriction. Is it worth exploring this option?

mo-abodas replied:
Yes, the current approach is not very elegant, I'll try and code it in a more flexible way.

mo-abodas commented:
@valeriupredoi probably a stupid question, but I don't understand why the tests are failing with the error:
No module named ESMF.
I guess I must have done something wrong, but I can't see what it is!

valeriupredoi commented:
@mo-abodas please merge `main` here and push so it picks up the latest devel main branch, that'll fjx the tests too 👍

valeriupredoi commented:
Long story is that since esmpy=8.4.0 the import of ESMF has changed to import esmpy (and a whole lot of other stuff), we fixed that in main dev branch a while ago, so you should be ok

alistairsellar reviewed (approved):
Thanks for the changes @mo-abodas! All good from a science review perspective.

bouweandela commented:
The deadline for including new work in [v2.9](https://github.com/ESMValGroup/ESMValTool/discussions/3198) has passed, so I'm removing this issue from the milestone. Please add it to the [next milestone](https://github.com/ESMValGroup/ESMValTool/milestone/13) if you plan to include this in the next release.

ehogan reviewed (changes requested):
Thanks @alejandrobodas! 😊 I ran your recipe at the MO and it worked for me :)

I have a few general comments for your consideration:

- Would you consider using more explicit variable names? There are many three letter variable names (e.g. efp, nnn, vvv, zzz, etc.) that do not mean anything to a technical person like me! Also, I would recommend using the more readable `point` rather than `pnt` (which I assume means `point`!) in favour of saving 2 characters (readability counts!) 😊 
- There are a number of calls to `print()` within the `single_model_diagnostics.py` diagnostic file; given you have initialised a logger, would it be worth using the logger instead?
- Would it be possible for you to expand your docstrings to include a description of the arguments, please? (There are example on how to do this in the [radiation budget recipe](https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/diag_scripts/radiation_budget/radiation_budget.py)).
- "The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.", see [PEP8](https://peps.python.org/pep-0008/#maximum-line-length). You can use something like Black to auto-format the diagnostic files to see how to do this (if you use `black` on the whole file, the ESMValTool pre-commit hooks will make changes / the checks will require some changes).
- I noticed there are a number of occurances of `cube.data`; running this command will realise the data. Is it possible to use `cube` instead to enable the benefits of Dask?
- There are a number of "magic numbers" used in the diagnostic files; would it be possible to clarify what they are, please? Adding a comment or a descriptive variable are a couple of examples of how to achieve this 😊 (I have commented on two examples)
- There are a number of `copy`s used in the diagnostic files; are they all necessary?

bouweandela commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+            'righi_mattia',
+        ],
+        'references': [
+            'acknow_project',
```
```suggestion
            'pearce23jclim',
```

mo-abodas replied:
I've simplified the provenance record by deleting that information, already available in the recipe's yml file.

bouweandela replied:
Thanks, that's fine. Just as an explanation of why we have the option of specifying it in the diagnostic script too: if someone else would write a different recipe that uses this diagnostic code, the reference would be lost.

mo-abodas replied:
Thanks for the explanation. In that case it makes sense to keep the reference to the paper. I've done this in aec22df9f6b0ef9aaa2d5e23c0d90cafd19ba69e.

bouweandela commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+        'domains': ['global'],
+        'plot_types': [plot_type],
+        'authors': [
+            'andela_bouwe',
+            'righi_mattia',
```
Replace with the actual authors?

mo-abodas replied:
I've simplified the provenance record by deleting that information, The correct authors are listed in the recipe's yml file.

ehogan commented on esmvaltool/diag_scripts/iht_toa/poisson_solver.py:
```diff
@@ -0,0 +1,285 @@
+"""(C) Crown Copyright 2023, the Met Office.
```
I received a response from our legal team and "The short answer is yes, you need to include the Met Office copyright notice in your modification". I will plan to add one to the radiation budget recipe Jon and I contributed! ;)

However, I would recommend adding the copyright as a comment, rather than in the module docstring.

mo-abodas replied:
Fixed in 53e7b08a8fe8c9b22547c242a71e60420fd9eda8

ehogan commented on doc/sphinx/source/recipes/recipe_iht_toa.rst:
```diff
@@ -0,0 +1,123 @@
+
+Recipes are stored in esmvaltool/recipes/
+
+    * recipe_iht_toa.yml calculates the IHT maps for the following radiative fluxes:
+
+        * Total net, SW net, LW net (Figure 2).
+        * Total CRE, SW CRE, LW CRE (Figure 4).
+        * All-sky and clear-sky reflected SW (Figure 5).
+        * The meridional heat transports (MHT) of the fluxes above (Figures 1 and 3).
```
I recognise that this is how it is formatted in the template; I have opened #3377 to address this :)

mo-abodas replied:
Changed as suggested, it looks OK now.

ehogan replied:
The second set of points are still showing as a quote at https://esmvaltool--3177.org.readthedocs.build/en/3177/recipes/recipe_iht_toa.html#available-recipes-and-diagnostics. Since I'm adding a suggestion that will fix this I will resolve this conversation :)

ehogan commented on doc/sphinx/source/recipes/recipe_iht_toa.rst:
```diff
@@ -0,0 +1,123 @@
+
+Diagnostics are stored in esmvaltool/diag_scripts/iht_toa/
+
+    * single_model_diagnostics.py: driver script that produces the plots.
+    * poisson_solver.py: solver that calculates the IHTs.
```
These bullet points have rendered as a quote, see above.

ehogan commented on doc/sphinx/source/recipes/recipe_iht_toa.rst:
```diff
@@ -0,0 +1,123 @@
+
+User settings in recipe
+-----------------------
+
```
If there are no user settings in the recipe, it is better to say "There are no user settings in this recipe" or equivalent, rather than leaving this section empty :)

mo-abodas replied:
Changed as suggested.

ehogan commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+
+def get_provenance_record(plot_type, ancestor_files):
+    """Create a provenance record describing the diagnostic data and plot."""
+    record = {
+        'statistics': ['mean'],
+        'domains': ['global'],
+        'plot_types': [plot_type],
+        'authors': [
+            'andela_bouwe',
+            'righi_mattia',
+        ],
+        'references': [
+            'acknow_project',
+        ],
```
There is a link to "Provenance information" as part of the PR checklist; within that, the [Provenance items provided by the diagnostic script](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#provenance-items-provided-by-the-diagnostic-script) section provides keys that can be used in the provenance dictionary; for example, the `caption` is used for the plots in the HTML output from the recipe, so specifying this will replace the path below with the caption:

![image](https://github.com/ESMValGroup/ESMValTool/assets/918142/21f704e6-440f-47ae-8fbc-b0611c96f1d0)

Also, be sure that the `ancestor_files` contain only the files used for each plot! :)

mo-abodas replied:
I've simplified the function get_provenance_record, added figure captions and fixed the ancestor files in b46c7924f08eb84d4397dee28bce147a5da6744e

ehogan reviewed (changes requested):
Hi @mo-abodas, I have added some additional comments 😊 

+ Is it possible to use an already existing Poisson solver, for example, the one in SciPy: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.poisson.html?
+ Would you like to be listed as an author of ESMValTool on [Zenodo](https://zenodo.org/record/8120970)? If so, there are instructions to do this available in the [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) link in the PR checklist 😊

bouweandela commented on esmvaltool/recipes/recipe_iht_toa.yml:
```diff
@@ -0,0 +1,119 @@
+datasets:
+  # The CERES-EBAF observations reproduce the results of Pearce and
+  # Bodas-Salcedo (2023), but it can be applied to model data.
+  - {dataset: CERES-EBAF, project: obs4MIPs, level: L3B, version: Ed2-7, tier: 1, start_year: 2001, end_year: 2010}
```
Would it be possible to replace this by the version of the dataset that is available on ESGF? See https://github.com/ESMValGroup/ESMValTool/issues/2974#issuecomment-1372136192 for more details.

mo-abodas replied:
I've changed the dataset to Ed2-8.

bouweandela replied:
Thanks! Please also remove the `version` and `level` facets. The current value of the `version` facet does not correspond to a version that is available on ESGF (on ESGF there is only `version: v20160610`).

mo-abodas replied:
Thanks for spotting this. Fixed in aec22df9f6b0ef9aaa2d5e23c0d90cafd19ba69e.

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/poisson_solver.py:
```diff
@@ -0,0 +1,285 @@
+
+            if iteration == max_iterations:
+                self.logger.error('Poisson solver has not converged.')
+                sys.exit(1)
```
Fixed in 53e7b08a8fe8c9b22547c242a71e60420fd9eda8

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+)
+
+# Initialise logger
+logger = logging.getLogger(Path(__file__).stem)
```
Fixed in 53e7b08a8fe8c9b22547c242a71e60420fd9eda8

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+
+def var_name_constraint(var_name):
+    """Shortcut to create constraint for variable name."""
+    return iris.Constraint(cube_func=lambda c: c.var_name == var_name)
```
Fixed in c16c4b29a56b85730cee267629676e5494005e05

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+                                    long_name=long_name,
+                                    var_name=var_name,
+                                    units="PW",
+                                    dim_coords_and_dims=[(time_coord, 0)])
```
It's a units conversion, I've now used a variable with an explanatory name for the conversion.

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+                           color='w')
+            format_plot(plt.gca(), label[i][0], title[i][0])
+
+            plt.subplot(grds[i * 7:(i * 7) + 7, 1],
```
It's the step used to separate subplots in the gridspec. I've replace the magic numbers with a variable.

mo-abodas commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,714 @@
+        legend_label = ["TOA net all-sky", "TOA net clear-sky"]
+
+        plt.figure(figsize=(6, 12))
+        for i, var_name in enumerate(var_list):
```
Done as suggested.

mo-abodas commented:
@ehogan @bouweandela thanks for your reviews! I believe I've addressed all your comments, please let me know if you think anything else is needed.

@ehogan , please see below specific responses to your general comments:

> Would you consider using more explicit variable names? There are many three letter variable names (e.g. efp, nnn, vvv, zzz, etc.) that do not mean anything to a technical person like me! Also, I would recommend using the more readable point rather than pnt (which I assume means point!) in favour of saving 2 characters (readability counts!)

I've done  this in commits 3a56318cc17662cc27fcac669df06f67fdfce00d, ee0a670aaa07836b226c9f2b1e51073f047712ef, 6306ff2820e8d12e759d995f1ae93d7cd44d7241. I've expanded most of the three-letter names with more meaningful names. However, there are a couple of exceptions where I think it's better to keep the original names. I've kept `efp` and `mht` as variable names and defined the acronyms in the docstring of the `ImpliedHeatTransport` class. In `poisson_solver.py`, I've kept the original names in the methods `SphericalPoisson.set_matrices` and `SphericalPoisson.solve`. The reason for this is that the short names are more traceable to the names in the paper that describes the numerical algorithm.

> There are a number of calls to print() within the single_model_diagnostics.py diagnostic file; given you have initialised a logger, would it be worth using the logger instead?

Done in 449e2e9d7cc97ebdd5c9522d1e890fe438b82357.

> Would it be possible for you to expand your docstrings to include a description of the arguments, please? (There are examples on how to do this in the radiation budget recipe).

Done in 4b44c60f778cbdadbe75f061d770f1dd4390b470, 417f0a445ce3b753b727e0b2598401652fac0ddf.

> "The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.", see PEP8. You can use something like Black to auto-format the diagnostic files to see how to do this (if you use black on the whole file, the ESMValTool pre-commit hooks will make changes / the checks will require some changes).

Done in 2f54ae3414db302910bfe195d4075d8f3526b9b4.

> I noticed there are a number of occurances of cube.data; running this command will realise the data. Is it possible to use cube instead to enable the benefits of Dask?

Done in f4f8c5821f201241049508e130f5d358ea01e0ae. Exceptions are calls to matplotlib plotting functions that require numpy arrays as arguments.

> There are a number of "magic numbers" used in the diagnostic files; would it be possible to clarify what they are, please? Adding a comment or a descriptive variable are a couple of examples of how to achieve this 😊 (I have commented on two examples)

Done in 000206c16e3a4efe1cbce571b6ce211c37781806, 015eec3f08f21181fc8bec4f9d2739c6a5180b9a, 40b6081da5dcb61fdd9d97618404a26fea5286fe.

> There are a number of copys used in the diagnostic files; are they all necessary?

Only the one in the poisson solver. I've replaced the others.
Done in 9c64e8ed239d7fc98719a8739bc79ae29f321203.

valeriupredoi reviewed (approved):
refreshing my stale review - my worries were just numba-related, all good, serious heavy-lifting review done by others :beer:

ehogan commented on CITATION.cff:
```diff
@@ -8,6 +8,10 @@ authors:
     family-names: Andela
     given-names: Bouwe
     orcid: "https://orcid.org/0000-0001-9005-8940"
+  -
+    affiliation: "MetOffice, UK"
+    family-names: Bodas-Salcedo
+    given-names: Alejandro
```
It looks like this top section might be for the core team; would it be possible to move this entry to the alphabetical list that starts lower down in this file (around ~L70), please? :)

ehogan commented on doc/sphinx/source/recipes/recipe_iht_toa.rst:
```diff
@@ -0,0 +1,123 @@
+
+* recipe_iht_toa.yml calculates the IHT maps for the following radiative fluxes:
+
+   * Total net, SW net, LW net (Figure 2).
+   * Total CRE, SW CRE, LW CRE (Figure 4).
+   * All-sky and clear-sky reflected SW (Figure 5).
+   * The meridional heat transports (MHT) of the fluxes above (Figures 1 and 3).
```
This is still showing as a quote at https://esmvaltool--3177.org.readthedocs.build/en/3177/recipes/recipe_iht_toa.html#available-recipes-and-diagnostics; the following should fix this :)

```suggestion
  * Total net, SW net, LW net (Figure 2).
  * Total CRE, SW CRE, LW CRE (Figure 4).
  * All-sky and clear-sky reflected SW (Figure 5).
  * The meridional heat transports (MHT) of the fluxes above (Figures 1 and 3).
```

ehogan commented:
> @ehogan @bouweandela thanks for your reviews! I believe I've addressed all your comments, please let me know if you think anything else is needed.

Nice work @alejandrobodas! 🥳 

> > There are a number of calls to print() within the single_model_diagnostics.py diagnostic file; given you have initialised a logger, would it be worth using the logger instead?
> 
> Done in [449e2e9](https://github.com/ESMValGroup/ESMValTool/commit/449e2e9d7cc97ebdd5c9522d1e890fe438b82357).

Should the `print` on L986 also be updated?

ehogan commented on esmvaltool/diag_scripts/iht_toa/single_model_diagnostics.py:
```diff
@@ -0,0 +1,994 @@
+
+
+def matching_strings(list_of_strings, substrings):
+    """Return the subset of strings that  of files .
```
This sentence looks like it is missing some words! :)

ehogan reviewed (changes requested):
Many thanks for all your hard work on addressing our review comments, @alejandrobodas! 🥳 I spotted a few minor things 😊

mo-abodas commented:
@ehogan thanks for spotting a few mistakes! I've fixed them as suggested. I've deleted the print statement because it didn't really provide any useful information.

ehogan reviewed (approved):
Thanks @alejandrobodas! 🥳

alistairsellar commented:
🥳 Well done @mo-abodas and everyone who helped with this!

mo-abodas commented:
Thanks everyone for your help!
